### PR TITLE
Screen 7 - Turn Off Secure Boot

### PR DESCRIPTION
**Pull request by charafau**
Implemented screen 7 design with model

image

image

image

issue: https://github.com/canonical/ubuntu-desktop-installer/issues/32

### CODE REVIEW

**Comment by charafau** [comment]
@oSoMoN  not sure what I need to change to make this merge (I signed cla before)

**Comment by Feichtmeier** [comment]
I think you need to see that your launchpad account with which you've signed the contributors agreement matches the github email address

Also all pages now descent from the wizardpage class, so maybe you need to adapt your PR now

**Comment by charafau** [comment]
@Feichtmeier  thanks! I got one page merged so I think it's just github check error (I have launchpad accoutn signed up). Let me guys know if you want to change this and use wizard page 😉

**Comment by oSoMoN** [comment]
It looks like the CLA check failed because of oSoMoN@users.noreply.github.com, because I used github's web UI to resolve trivial merge conflicts. No need to worry about this.

If you don't mind rebasing on the WizardPage, that'd be great!

**Comment by charafau** [comment]
@oSoMoN  sorry for not replying. I rebased with master and move code to wizard widgets. CI still fails but tests pass for me locally. Is there anything I should do?

**Comment by Feichtmeier** [comment]
> @oSoMoN sorry for not replying. I rebased with master and move code to wizard widgets. CI still fails but tests pass for me locally. Is there anything I should do?

The cla-check fails because @oSoMoN force pushed into your branch and the names do not match (I think)

Ht codecov CI checks how much of the code is covered by tests before and after your PR and if it is less the CI fails. So you need to add a test for this page in the test dir I think :)

**Comment by oSoMoN** [comment]
> The cla-check fails because @oSoMoN force pushed into your branch and the names do not match (I think)

I used github's web editor to resolve merge conflicts, and the CLA check doesn't like my anonymized github handle. You can safely ignore this.

File: packages/ubuntu_desktop_installer/lib/pages/turn_off_secure_boot/turn_off_secure_boot_page.dart [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+                    },
+                    decoration: InputDecoration(
+                        border: const OutlineInputBorder(),
+                        hintText: lang.chooseSecurityKey,
+                        enabled: model.areTextFieldEnabled),
+                  ),
+                ),
+                const SizedBox(height: _kFormElementPadding),
+                _PasswordSizedBox(
+                  child: TextFormField(
```
Comment (jpnurmi) [inline packages/ubuntu_desktop_installer/lib/pages/turn_off_secure_boot/turn_off_secure_boot_page.dart:65]: could you use `ValidatedFormField` similarly to the _Who are you_ page?

File: packages/ubuntu_desktop_installer/lib/pages/turn_off_secure_boot/turn_off_secure_boot_page.dart [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+                  onChanged: model.setSecureBootMode,
+                ),
+              ],
+            ),
+          ),
+        ),
+        actions: <WizardAction>[
+          WizardAction(
+            label: lang.backButtonText,
+            onActivated: Navigator.of(context).pop,
```
Comment (jpnurmi) [inline packages/ubuntu_desktop_installer/lib/pages/turn_off_secure_boot/turn_off_secure_boot_page.dart:98]: `Wizard.of(context).back`

File: packages/ubuntu_desktop_installer/lib/pages/turn_off_secure_boot/turn_off_secure_boot_page.dart [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+          ),
+          WizardAction(
+            label: lang.continueButtonText,
+            onActivated: model.isFormValid
```
Comment (jpnurmi) [inline packages/ubuntu_desktop_installer/lib/pages/turn_off_secure_boot/turn_off_secure_boot_page.dart:102]: `Wizard.of(context).next`, and feel free to use `enabled: model.isFormValid` to get rid of the condition in the callback, to make the code look (subjectively) nicer

**Comment by charafau** [comment]
@jpnurmi  took me a while but should be ok, let me know if you want to change something

File: packages/ubuntu_desktop_installer/lib/l10n/app_en_US.arb [hunk 4]
```diff
@@ -48,6 +48,16 @@
   "pressOneKey": "Please press one of the following keys:",
   "isKeyPresent": "Is the following key present on your keyboard?",
 
+  "turnOffSecureBootTitle": "Turn off secure boot",
+  "turnOffSecureBootDescription": "You've chosen to install third-party driver software. This requires turning off  Secure Boot.\nTo do this, you need to choose a security key now, and enter it when the system restarts.",
```
Comment (jpnurmi) [inline packages/ubuntu_desktop_installer/lib/l10n/app_en_US.arb:52]: extra space before "Secure Boot"

File: packages/ubuntu_desktop_installer/lib/pages/turn_off_secure_boot/turn_off_secure_boot_model.dart [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    _mode = mode;
+
+    _validateForm();
+
+    notifyListeners();
+  }
+
+  void setSecurityKey(String? key) {
```
Comment (jpnurmi) [inline packages/ubuntu_desktop_installer/lib/pages/turn_off_secure_boot/turn_off_secure_boot_model.dart:39]: could be non-nullable?

File: packages/ubuntu_desktop_installer/lib/pages/turn_off_secure_boot/turn_off_secure_boot_model.dart [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    _securityKey = key;
+
+    _validateForm();
+
+    notifyListeners();
+  }
+
+  void setConfirmKey(String? key) {
```
Comment (jpnurmi) [inline packages/ubuntu_desktop_installer/lib/pages/turn_off_secure_boot/turn_off_secure_boot_model.dart:51]: same, could be non-nullable?

File: packages/ubuntu_desktop_installer/lib/pages/turn_off_secure_boot/turn_off_secure_boot_page.dart [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+  final _formKey = GlobalKey<FormState>();
+
+  @override
+  Widget build(BuildContext context) {
+    final model = context.watch<TurnOffSecureBootModel>();
+    return LocalizedView(
+      builder: (context, lang) => WizardPage(
+        title: Text(lang.turnOffSecureBootTitle),
+        content: Container(
+          padding: const EdgeInsets.symmetric(horizontal: 20),
```
Comment (jpnurmi) [inline packages/ubuntu_desktop_installer/lib/pages/turn_off_secure_boot/turn_off_secure_boot_page.dart:34]: remove the container, WizardPage defines the "standard" padding

File: packages/ubuntu_desktop_installer/lib/pages/turn_off_secure_boot/turn_off_secure_boot_page.dart [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+          child: Form(
+            key: _formKey,
+            child: Column(
+              crossAxisAlignment: CrossAxisAlignment.stretch,
+              children: [
+                Align(
+                  alignment: Alignment.centerLeft,
+                  child: Text(lang.turnOffSecureBootDescription),
```
Comment (jpnurmi) [inline packages/ubuntu_desktop_installer/lib/pages/turn_off_secure_boot/turn_off_secure_boot_page.dart:42]: move as page header:

```dart
WizardPage(
    title: Text(lang.turnOffSecureBootTitle),
    header: Text(lang.turnOffSecureBootDescription),
    // ...
),
```

File: packages/ubuntu_desktop_installer/lib/pages/turn_off_secure_boot/turn_off_secure_boot_page.dart [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+                ),
+                RadioListTile<SecureBootMode>(
```
Comment (jpnurmi) [inline packages/ubuntu_desktop_installer/lib/pages/turn_off_secure_boot/turn_off_secure_boot_page.dart:44]: s/RadioListTile/RadioButton

File: packages/ubuntu_desktop_installer/lib/pages/turn_off_secure_boot/turn_off_secure_boot_page.dart [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+                ),
+                const SizedBox(height: 24),
+                PasswordConfirmFormField(
+                  onChanged: (value) {
+                    model.setSecurityKey(value);
+                    _formKey.currentState?.validate();
+                  },
+                ),
+                const SizedBox(height: _kFormElementPadding),
+                RadioListTile<SecureBootMode>(
```
Comment (jpnurmi) [inline packages/ubuntu_desktop_installer/lib/pages/turn_off_secure_boot/turn_off_secure_boot_page.dart:65]: s/RadioListTile/RadioButton

File: packages/ubuntu_desktop_installer/lib/pages/turn_off_secure_boot/turn_off_secure_boot_page.dart [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+                  onChanged: model.setSecureBootMode,
+                ),
+                const SizedBox(height: _kFormElementPadding),
+                PasswordFormField(
+                  onChanged: (value) {
+                    model.setSecurityKey(value);
+                    _formKey.currentState?.validate();
+                  },
+                ),
+                const SizedBox(height: 24),
```
Comment (jpnurmi) [inline packages/ubuntu_desktop_installer/lib/pages/turn_off_secure_boot/turn_off_secure_boot_page.dart:57]: kContentSpacing

File: packages/ubuntu_desktop_installer/lib/pages/turn_off_secure_boot/turn_off_secure_boot_page.dart [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+                PasswordConfirmFormField(
+                  onChanged: (value) {
+                    model.setSecurityKey(value);
+                    _formKey.currentState?.validate();
+                  },
+                ),
+                const SizedBox(height: _kFormElementPadding),
```
Comment (jpnurmi) [inline packages/ubuntu_desktop_installer/lib/pages/turn_off_secure_boot/turn_off_secure_boot_page.dart:64]: kContentSpacing

File: packages/ubuntu_desktop_installer/lib/pages/turn_off_secure_boot/turn_off_secure_boot_widgets.dart [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+  final void Function(String) onChanged;
+
+  @override
+  Widget build(BuildContext context) {
+    final model = context.watch<TurnOffSecureBootModel>();
+    return FractionallySizedBox(
+      widthFactor: 0.80,
+      child: LocalizedView(builder: (context, lang) {
+        return ValidatedFormField(
+          spacing: 190,
```
Comment (jpnurmi) [inline packages/ubuntu_desktop_installer/lib/pages/turn_off_secure_boot/turn_off_secure_boot_widgets.dart:24]: this is the spacing between the field and its "checkmark" icon. i think we should pass a calculated fieldWith instead :)

File: packages/ubuntu_desktop_installer/lib/pages/turn_off_secure_boot/turn_off_secure_boot_widgets.dart [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+          obscureText: true,
+          labelText: lang.chooseSecurityKey,
+          onChanged: model.setSecurityKey,
+          enabled: model.areTextFieldEnabled,
+          validator: RequiredValidator(errorText: ''),
```
Comment (jpnurmi) [inline packages/ubuntu_desktop_installer/lib/pages/turn_off_secure_boot/turn_off_secure_boot_widgets.dart:29]: missing text?

File: packages/ubuntu_desktop_installer/lib/pages/turn_off_secure_boot/turn_off_secure_boot_widgets.dart [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+  @override
+  Widget build(BuildContext context) {
+    final model = context.watch<TurnOffSecureBootModel>();
+    return FractionallySizedBox(
+      widthFactor: 0.80,
+      child: LocalizedView(builder: (context, lang) {
+        return ValidatedFormField(
+          spacing: 190,
+          obscureText: true,
+          labelText: lang.chooseSecurityKey,
```
Comment (jpnurmi) [inline packages/ubuntu_desktop_installer/lib/pages/turn_off_secure_boot/turn_off_secure_boot_widgets.dart:26]: missing `successWidget: const SuccessIcon()`

File: packages/ubuntu_desktop_installer/lib/l10n/app_en_US.arb [hunk 16]
```diff
@@ -48,6 +48,16 @@
   "pressOneKey": "Please press one of the following keys:",
   "isKeyPresent": "Is the following key present on your keyboard?",
 
+  "turnOffSecureBootTitle": "Turn off secure boot",
```
Comment (jpnurmi) [inline packages/ubuntu_desktop_installer/lib/l10n/app_en_US.arb:51]: sorry, the title is wrong in the screenshot. in the [spec](https://people.canonical.com/~platform/desktop/installer/7.%20Turn%20Off%20Secure%20Boot/) it says "Configure secure boot" which is the correct one (following what ubiquity does, and this is more complex than flipping a switch anyway)

File: packages/ubuntu_desktop_installer/lib/l10n/app_en_US.arb [hunk 17]
```diff
@@ -48,6 +48,16 @@
...
+  "turnOffSecureBootDescription": "You've chosen to install third-party driver software. This requires turning off  Secure Boot.\nTo do this, you need to choose a security key now, and enter it when the system restarts.",
+  "turnOffSecureBootOption": "Turn off secure boot",
+  "chooseSecurityKey": "Choose a security key",
+  "confirmSecurityKey": "Confirm security key",
+  "dontInstallDriverSoftwareNow": "Don't install the driver software for now",
+  "dontInstallDriverSoftwareNowDescription": "You can install it later from Software & Updates Settings.",
+  "turnOffSecureBootPasswordRequired": "Password is required",
+  "secureBootPasswordsDontMatch": "Passwords do not match",
```
Comment (jpnurmi) [inline packages/ubuntu_desktop_installer/lib/l10n/app_en_US.arb:59]: should it be "security key"?

File: packages/ubuntu_desktop_installer/lib/pages/turn_off_secure_boot/turn_off_secure_boot_widgets.dart [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+
+  final void Function(String) onChanged;
+
+  @override
+  Widget build(BuildContext context) {
+    final model = context.watch<TurnOffSecureBootModel>();
+    return FractionallySizedBox(
+      widthFactor: 0.80,
+      child: LocalizedView(builder: (context, lang) {
+        return ValidatedFormField(
```
Comment (jpnurmi) [inline packages/ubuntu_desktop_installer/lib/pages/turn_off_secure_boot/turn_off_secure_boot_widgets.dart:50]: same as above applies to this

**Review by jpnurmi — COMMENTED** [review]
Thanks! Looks good - only a few minor things and then it's ready to go.

**Comment by jpnurmi** [comment]
Hi @charafau, open PRs get quickly outdated since the project is under heavy development. Let me know if you'd like me to adapt this to the latest changes.

**Comment by charafau** [comment]
@jpnurmi  sorry, dont have much time lately, rebased and fixed.

**Review by jpnurmi — APPROVED** [review]
Thanks! I have applied some last minor fixes so we can have this merged and avoid the risk of conflicts again!

**Pull request merged**